xla/hlo/analysis/hlo_replication_analysis.cc

```diff
@@ -94,45 +105,92 @@ HloReplicationAnalysis::DetermineHloInstructionIsReplicated(
         global_id = Cast<HloAllGatherInstruction>(hlo)->use_global_device_ids();
       }
       if (global_id) {
-        // TODO(philipphack): The following is incorrect if partitions are
-        // replicated differently on replicas, or if replicas are replicated
-        // differently on partitions.
-        bool replicated_across_partitions = true;
-        bool replicated_across_replicas = true;
         const int64_t num_partitions =
             hlo->GetModule()->config().num_partitions();
-        absl::flat_hash_set<int64_t> visited_partitions;
-        absl::flat_hash_set<int64_t> visited_replicas;
-        std::vector<int64_t> device_set;
-        std::vector<absl::Span<const int64_t>> device_sets;
-        std::vector<std::vector<int64_t>> device_sets_storage;
+        const int64_t replica_count =
+            hlo->GetModule()->config().replica_count();
+        // Holds the partition IDs of all replica groups in which a given
+        // replica participates.
+        std::vector<std::vector<absl::flat_hash_set<int64_t>>>
+            partitions_for_replica(replica_count);
+        // Holds the replica IDs of all replica groups in which a given
+        // partition participates.
+        std::vector<std::vector<absl::flat_hash_set<int64_t>>>
+            replicas_for_partition(num_partitions);
+        std::vector<std::vector<int64_t>> device_sets;
         for (const auto& group : hlo->replica_groups()) {
-          device_set.clear();
-          visited_partitions.clear();
-          visited_replicas.clear();
-          visited_replicas.reserve(group.replica_ids().size());
-          visited_partitions.reserve(group.replica_ids().size());
+          std::vector<int64_t> device_set;
+          std::vector<IDs> pids_and_rids;
+          // Convert the global IDs in the replica group to partition and
+          // replica IDs.
           for (int64_t id : group.replica_ids()) {
             int64_t rid = id / num_partitions;
             int64_t pid = id % num_partitions;
-            visited_partitions.insert(pid);
-            visited_replicas.insert(rid);
+            pids_and_rids.emplace_back(IDs{pid, rid});
             if (support_partial_replication) {
               device_set.push_back(cross_partition_spmd ? pid : rid);
             }
           }
-          replicated_across_partitions &=
-              visited_partitions.size() == num_partitions;
-          replicated_across_replicas &=
-              visited_replicas.size() ==
-              hlo->GetModule()->config().replica_count();
           if (support_partial_replication) {
-            device_sets_storage.push_back(device_set);
-            device_sets.push_back(device_sets_storage.back());
+            device_sets.push_back(std::move(device_set));
```

Comment:
This almost works, but it will potentially fail in the case where cross_partition_spmd=true if a device_sets contains different replicas. For example, I think the following case fails by having HloInstructionIsReplicatedAt incorrectly return true.

num_partitions = 2
num_replicas = 2
cross_partition_spmd = true
all gather replica_groups = {{0,3}{1,2}}
replica_groups passed to HloInstructionIsReplicatedAt = {{0,1}}

I haven't personally tried this case, but the issue is that device_sets will be {{0,1},{0,1}}. It will seem replicated since the two sets each have both partitions: 0 and 1. But they are on different replicas, and so are not truly replicated.

---

I'll describe the algorithm I think you should implement. @pkanwar23 helped me come up with it. I will use this example when describing it:

num_partitions = 6
num_replicas = 2
cross_partition_spmd = true
all gather replica_groups = {{0,1,4,5,6,10},{2,3,7,8,9,11}

The algorithm is as follows. Let `M = num_partitions, N = num_replicas`. In DetermineHloInstructionIsReplicated, create an array `int id_to_group[M][N]`, such that `id_to_group[m][n]` maps device `(m, n)` to the replica_group that device appears in. For example, the array in the example above would be (I use A to refer to replica_group 0 and B to refer to replica_group 1):

| | replica 0  | replica 1 |
| --- | ------------- | ------------- |
| **partition 0** | A  | A  |
| **partition 1** | A  | B  |
| **partition 2** | B  | B  |
| **partition 3** | B  | B  |
| **partition 4** | A  | A  |
| **partition 5** | A  | B  |

Store `id_to_group`  inside PartiallyReplicated and return it from `DetermineHloInstructionIsReplicated`

Within `IsReplicatedWithinSubgroup`, we are given a `replica_groups` parameter. We must use this and the `id_to_group` table to determine if for each replica, the partitions are replicated according to `replica_groups`. I'll describe how with a concrete example. Suppose the `replica_groups` parameter is `{0,4},{1,5},{2,3}`. We process each of the three groups separately, returning false if any group is non-replicated. For `{0,4}`, we check that rows 0 and 4 in the table are identical. In this case, they are both `(A, A)`, so we do not return false. Similarly, rows 1 and 5 are the same (both `(A, B)`), and rows 2 and 3 are the same (both `(B, B)`). So all groups are replicated, and we return true.

Now suppose `replica_groups` is instead `{0,4,5},{1,2,3}`. In this case, rows 4 and 5 within the first group are not the same , so we return false.

Reply:
Thanks for the suggestion! As discussed, I made a small change to the existing implementation which should cover the case you mentioned.